src/box/lua/net_box.lua

```diff
@@ -758,6 +758,9 @@ end
 
 function remote_methods:ping(opts)
     check_remote_arg(self, 'ping')
+    if opts ~= nil and type(opts) ~= 'table' then
+        box.error(E_PROC_LUA, 'opts must be a table')
```

Comment:
The issue is unfortunately broader:
```
tarantool> c:ping({on_push = 1, is_async = 1})
---
- false
...
```
It happens for any exception, not just for invalid index access like when opts is not nil and is not a table either.

I think a proper fix should
1. Add validation of opts type into `remote_methods:_request`;
2. Remove `pcall` from `remote_methods:ping()` and remove `box.error()` calls from `remote_methods:_request()`. Add the `box.error()` to where `_request()` is used;
3. Then `:ping()` can let exceptions through. And return false when `:_request()` returns `nil, err`.

Slightly alternative approach could be to make `remote_methods:_request()` not call `box.error()` + introduce `remote_methods:_request_ex()` which calls `:_request()` and calls `box.error()` when gets `nil, err`. `:_request_ex()` would be used everywhere except `:ping()` which will use `:_request()` directly.

Moreover, because of `pcall()` the option `is_async` doesn't work with `:ping()` at all. It will just return true without the actual future object. This would be fixed alongside.

Reply:
I suppose that the expected behavior of the `remote_methods:ping()` function is to return only true or false according to the state of the server and to raise an error if user is doing something wrong or if an internal error has occurred. 

The solution you proposed (returning false in case `remote_methods:_request()` returns `nil, err`) doesn't solve the issue as the behavior of the `:ping()` remains the same as using `pcall` (if I understood you correctly). I think we should process the returned error and decide if it's the connection related one (return false from ping is such case) or another one (raise an error). Moreover there's no need to remove `pcall` as it lets exceptions go through. We just need to process them. 

Speaking of `is_async` option, it's not fixed alongside. We indeed can return future object in ping, but it's not very representative as on success `res` in `:_request` is `nil`. Moreover, `is_async` option is not supported according to docs. If it's really needed I can think of something about this.

Reply:
> The solution you proposed (returning false in case remote_methods:_request() returns nil, err) doesn't solve the issue as the behavior of the :ping() remains the same as using pcall (if I understood you correctly)

What is wrong with that? This is the expected behaviour - return true or false if the parameters were fine and there are no usage errors.

> I think we should process the returned error and decide if it's the connection related one (return false from ping is such case) or another one (raise an error). 

The problem here is that the number of connection-related codes is not fixed in the stone forever. New ones tend to be added and old ones can be dropped. For example, in your new patch you missed `ER_SYSTEM` which can be raised by any IO operation (all places where `SocketError` is raised in C code). You will raise an exception even though it should have beed `false`. I would advice against checking precise codes.

> Speaking of is_async option, it's not fixed alongside. We indeed can return future object in ping, but it's not very representative as on success res in :_request is nil.

I don't think I follow. Why wouldn't it be fixed? It works pretty well with all the other request types.

> Moreover, is_async option is not supported according to docs.

If something is not supported, its support can be added. We can't drop support of a feature, but we can add new. Although here it is rather a bug that a feature request. Using `is_async` I not only don't get any error, but I also can get `true` even when the server is unresponsive. Here is a short test:
```Lua
--
-- Instance 1
--
-- Step 1
--
box.cfg{listen = 3313}

-- Step 3
--
while true do end

--
-- Instance 2
--
-- Step 2
--
net = require('net.box')
c = net.connect(3313)

-- Step 4
--
c:ping({is_async = true})
```
The ping on step 4 returns `true` without even going to the server. When I remove `is_async`, the ping hangs, like it should.

I still would suggest to fix the ping properly while we are at it.